<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><link rel="canonical" href="https://www.jkobject.com/benGRN/base/" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>base - bengrn</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="../assets/_mkdocstrings.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "base";
        var mkdocs_page_input_path = "base.md";
        var mkdocs_page_url = "/benGRN/base/";
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> bengrn
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="" href="../notebooks/example.ipynb">example notebook</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">documentation</span></p>
              <ul class="current">
                  <li class="toctree-l1 current"><a class="reference internal current" href="#">base</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#bengrn.base">base</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#bengrn.base.BenGRN">BenGRN</a>
        <ul>
    <li class="toctree-l4"><a class="reference internal" href="#bengrn.base.BenGRN.compare_to">compare_to</a>
    </li>
    <li class="toctree-l4"><a class="reference internal" href="#bengrn.base.BenGRN.scprint_benchmark">scprint_benchmark</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#bengrn.base.compute_epr">compute_epr</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#bengrn.base.compute_genie3">compute_genie3</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#bengrn.base.compute_pr">compute_pr</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#bengrn.base.compute_scenic">compute_scenic</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#bengrn.base.get_GT_db">get_GT_db</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#bengrn.base.get_perturb_gt">get_perturb_gt</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#bengrn.base.get_scenicplus">get_scenicplus</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#bengrn.base.get_sroy_gt">get_sroy_gt</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#bengrn.base.load_genes">load_genes</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#bengrn.base.precision_recall">precision_recall</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#bengrn.base.train_classifier">train_classifier</a>
    </li>
        </ul>
    </li>
    </ul>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">bengrn</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
          <li class="breadcrumb-item">documentation</li>
      <li class="breadcrumb-item active">base</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="documentation-for-base-module">Documentation for <code>base</code> module</h1>


<div class="doc doc-object doc-module">



<h2 id="bengrn.base" class="doc doc-heading">
            <code>bengrn.base</code>


</h2>

    <div class="doc doc-contents first">

      <p>bengrn base module.</p>



  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h3 id="bengrn.base.BenGRN" class="doc doc-heading">
              <code class="highlight language-python"><span class="n">BenGRN</span></code>

</h3>


    <div class="doc doc-contents ">


      <p>Initializes the BenGRN class.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>grn</code></b>
                  (<code><span title="grnndata.GRNAnnData">GRNAnnData</span></code>)
              –
              <div class="doc-md-description">
                <p>The Gene Regulatory Network data.</p>
              </div>
            </li>
            <li>
              <b><code>full_dataset</code></b>
                  (<code><span title="typing.Optional">Optional</span>[<span title="anndata.AnnData">AnnData</span>]</code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>The full dataset, defaults to None.</p>
              </div>
            </li>
            <li>
              <b><code>doplot</code></b>
                  (<code>bool</code>, default:
                      <code>True</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to plot the results, defaults to True.</p>
              </div>
            </li>
            <li>
              <b><code>do_auc</code></b>
                  (<code>bool</code>, default:
                      <code>True</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to calculate the Area Under the Precision-Recall Curve, defaults to True.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
                  <details class="quote">
                    <summary>Source code in <code>bengrn/base.py</code></summary>
                    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">grn</span><span class="p">:</span> <span class="n">GRNAnnData</span><span class="p">,</span>
    <span class="n">full_dataset</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">AnnData</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">doplot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">do_auc</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Initializes the BenGRN class.</span>

<span class="sd">    Args:</span>
<span class="sd">        grn (GRNAnnData): The Gene Regulatory Network data.</span>
<span class="sd">        full_dataset (Optional[AnnData]): The full dataset, defaults to None.</span>
<span class="sd">        doplot (bool): Whether to plot the results, defaults to True.</span>
<span class="sd">        do_auc (bool): Whether to calculate the Area Under the Precision-Recall Curve, defaults to True.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">grn</span> <span class="o">=</span> <span class="n">grn</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">full_dataset</span> <span class="o">=</span> <span class="n">full_dataset</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">doplot</span> <span class="o">=</span> <span class="n">doplot</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">do_auc</span> <span class="o">=</span> <span class="n">do_auc</span>
</code></pre></div></td></tr></table></div>
                  </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h4 id="bengrn.base.BenGRN.compare_to" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compare_to</span></code>

</h4>


    <div class="doc doc-contents ">

      <p>compare_to compares the GRN to another GRN.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>other</code></b>
                  (<code><span title="typing.Optional">Optional</span>[<span title="grnndata.GRNAnnData">GRNAnnData</span>]</code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>The other GRN to compare to. Defaults to None.
If not given can use a default GRN from the 'to' argument.</p>
              </div>
            </li>
            <li>
              <b><code>to</code></b>
                  (<code>str</code>, default:
                      <code>&#39;collectri&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The name of the other GRN to compare to. Defaults to "collectri".
If 'other' is given, this argument is ignored.</p>
              </div>
            </li>
            <li>
              <b><code>organism</code></b>
                  (<code>str</code>, default:
                      <code>&#39;human&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The organism of the GRN to compare to. Defaults to "human".</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>dict</code></b>              –
              <div class="doc-md-description">
                <p>The metrics of the comparison.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>bengrn/base.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">compare_to</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">other</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">GRNAnnData</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">to</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;collectri&quot;</span><span class="p">,</span>
    <span class="n">organism</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;human&quot;</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    compare_to compares the GRN to another GRN.</span>

<span class="sd">    Args:</span>
<span class="sd">        other (Optional[GRNAnnData], optional): The other GRN to compare to. Defaults to None.</span>
<span class="sd">            If not given can use a default GRN from the &#39;to&#39; argument.</span>
<span class="sd">        to (str, optional): The name of the other GRN to compare to. Defaults to &quot;collectri&quot;.</span>
<span class="sd">            If &#39;other&#39; is given, this argument is ignored.</span>
<span class="sd">        organism (str, optional): The organism of the GRN to compare to. Defaults to &quot;human&quot;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: The metrics of the comparison.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">other</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">doplot</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;loading GT, &quot;</span><span class="p">,</span> <span class="n">to</span><span class="p">)</span>
        <span class="n">gt</span> <span class="o">=</span> <span class="n">get_GT_db</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">to</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="n">organism</span><span class="p">)</span>
        <span class="c1"># gt = gt[gt.type != &quot;post_translational&quot;]</span>
        <span class="n">varnames</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">gt</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">())</span>
        <span class="n">intersection</span> <span class="o">=</span> <span class="n">varnames</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">grn</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
        <span class="n">loc</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">grn</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">intersection</span><span class="p">)</span>
        <span class="n">adj</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">grn</span><span class="o">.</span><span class="n">varp</span><span class="p">[</span><span class="s2">&quot;GRN&quot;</span><span class="p">][:,</span> <span class="n">loc</span><span class="p">][</span><span class="n">loc</span><span class="p">,</span> <span class="p">:]</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">grn</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">loc</span><span class="p">,</span> <span class="s2">&quot;symbol&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

        <span class="n">da</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">adj</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">gt</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">genes</span> <span class="ow">and</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
                <span class="n">da</span><span class="p">[</span><span class="n">genes</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">i</span><span class="p">),</span> <span class="n">genes</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">j</span><span class="p">)]</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">doplot</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;intersection of </span><span class="si">{}</span><span class="s2"> genes&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">intersection</span><span class="p">)))</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;intersection pct:&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">intersection</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">grn</span><span class="o">.</span><span class="n">grn</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">elems</span> <span class="o">=</span> <span class="n">other</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">other</span><span class="o">.</span><span class="n">grn</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="n">da</span> <span class="o">=</span> <span class="n">other</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">grn</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">elems</span><span class="p">)</span><span class="o">.</span><span class="n">targets</span>
        <span class="k">if</span> <span class="n">da</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">5</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;da is very small: &quot;</span><span class="p">,</span> <span class="n">da</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1"># da = da.iloc[6:]</span>
        <span class="n">adj</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">grn</span><span class="o">.</span><span class="n">grn</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">da</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">da</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="n">da</span> <span class="o">=</span> <span class="n">da</span><span class="o">.</span><span class="n">values</span>

    <span class="k">return</span> <span class="n">compute_pr</span><span class="p">(</span>
        <span class="n">adj</span><span class="p">,</span>
        <span class="n">da</span><span class="p">,</span>
        <span class="n">doplot</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">doplot</span><span class="p">,</span>
        <span class="n">do_auc</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">do_auc</span><span class="p">,</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="bengrn.base.BenGRN.scprint_benchmark" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">scprint_benchmark</span></code>

</h4>


    <div class="doc doc-contents ">

      <p>scprint_benchmark full benchmarks of the GRN as in the scPRINT paper.</p>
<p>It will apply first an enrichment analysis over the [elems] of the GRN looking for TF enrichment and cell type marker gene enrichment
It will then apply an enrichment over each TF in the GRN for their targets in ENCODE.
Finaly, it will compare it to the OmniPath database GRN using precision recall type metrics.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>elems</code></b>
                  (<code>list</code>, default:
                      <code>[&#39;Central&#39;, &#39;Regulators&#39;, &#39;Targets&#39;]</code>
)
              –
              <div class="doc-md-description">
                <p>The genes in the GRN, to benchmark. Defaults to ["Central", "Regulators", "Targets"].
It corresponds to different views of the network.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>dict</code></b>              –
              <div class="doc-md-description">
                <p>The metrics of the benchmark.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>bengrn/base.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">scprint_benchmark</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">elems</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Central&quot;</span><span class="p">,</span> <span class="s2">&quot;Regulators&quot;</span><span class="p">,</span> <span class="s2">&quot;Targets&quot;</span><span class="p">]):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    scprint_benchmark full benchmarks of the GRN as in the scPRINT paper.</span>

<span class="sd">    It will apply first an enrichment analysis over the [elems] of the GRN looking for TF enrichment and cell type marker gene enrichment</span>
<span class="sd">    It will then apply an enrichment over each TF in the GRN for their targets in ENCODE.</span>
<span class="sd">    Finaly, it will compare it to the OmniPath database GRN using precision recall type metrics.</span>

<span class="sd">    Args:</span>
<span class="sd">        elems (list, optional): The genes in the GRN, to benchmark. Defaults to [&quot;Central&quot;, &quot;Regulators&quot;, &quot;Targets&quot;].</span>
<span class="sd">            It corresponds to different views of the network.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: The metrics of the benchmark.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;base enrichment&quot;</span><span class="p">)</span>
    <span class="n">metrics</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">elem</span> <span class="ow">in</span> <span class="n">elems</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">elem</span> <span class="o">==</span> <span class="s2">&quot;Central&quot;</span> <span class="ow">and</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">grn</span><span class="o">.</span><span class="n">varp</span><span class="p">[</span><span class="s2">&quot;GRN&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">100_000_000</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;too many genes for central computation&quot;</span><span class="p">)</span>
            <span class="k">continue</span>
        <span class="n">res</span> <span class="o">=</span> <span class="n">utils</span><span class="o">.</span><span class="n">enrichment</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">grn</span><span class="p">,</span>
            <span class="n">of</span><span class="o">=</span><span class="n">elem</span><span class="p">,</span>
            <span class="n">gene_sets</span><span class="o">=</span><span class="p">[</span>
                <span class="p">{</span><span class="s2">&quot;TFs&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">utils</span><span class="o">.</span><span class="n">TF</span><span class="p">]},</span>
                <span class="n">utils</span><span class="o">.</span><span class="n">file_dir</span> <span class="o">+</span> <span class="s2">&quot;/celltype.gmt&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="n">doplot</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">maxsize</span><span class="o">=</span><span class="mi">2000</span><span class="p">,</span>
            <span class="n">top_k</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">res</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="nb">len</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">res2d</span><span class="p">[(</span><span class="n">res</span><span class="o">.</span><span class="n">res2d</span><span class="p">[</span><span class="s2">&quot;FDR q-val&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.1</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">res2d</span><span class="p">[</span><span class="s2">&quot;NES&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)])</span>
            <span class="o">&gt;</span> <span class="mi">0</span>
        <span class="p">):</span>
            <span class="n">metrics</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="s2">&quot;enriched_terms_&quot;</span>
                    <span class="o">+</span> <span class="n">elem</span><span class="p">:</span> <span class="n">res</span><span class="o">.</span><span class="n">res2d</span><span class="p">[</span>
                        <span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">res2d</span><span class="p">[</span><span class="s2">&quot;FDR q-val&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.1</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">res2d</span><span class="p">[</span><span class="s2">&quot;NES&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span>
                    <span class="p">]</span><span class="o">.</span><span class="n">Term</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
                <span class="p">}</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">doplot</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">_</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">terms</span><span class="o">=</span><span class="s2">&quot;0__TFs&quot;</span><span class="p">)</span>
                <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="k">pass</span>
        <span class="n">istrue</span> <span class="o">=</span> <span class="n">metrics</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;TF_enr&quot;</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
        <span class="n">istrue</span> <span class="o">=</span> <span class="n">istrue</span> <span class="ow">or</span> <span class="p">(</span>
            <span class="n">res</span><span class="o">.</span><span class="n">res2d</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">res</span><span class="o">.</span><span class="n">res2d</span><span class="o">.</span><span class="n">Term</span> <span class="o">==</span> <span class="s2">&quot;0__TFs&quot;</span><span class="p">,</span> <span class="s2">&quot;FDR q-val&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.1</span>
        <span class="p">)</span>
        <span class="n">metrics</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;TF_enr&quot;</span><span class="p">:</span> <span class="n">istrue</span><span class="p">})</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">doplot</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;_________________________________________&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;TF specific enrichment&quot;</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">FILEDIR</span> <span class="o">+</span> <span class="s2">&quot;/../data/tfchip_data.json&quot;</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>
        <span class="n">tfchip</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
    <span class="n">TFinchip</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span><span class="p">:</span> <span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">tfchip</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>
    <span class="n">res</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span>
    <span class="n">previous_level</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">root</span><span class="o">.</span><span class="n">manager</span><span class="o">.</span><span class="n">disable</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">disable</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">WARNING</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">TFinchip</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">v</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">grn</span><span class="o">.</span><span class="n">grn</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="c1"># print(k)</span>
        <span class="n">j</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="n">test</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">grn</span><span class="o">.</span><span class="n">grn</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">loc</span><span class="p">[[</span><span class="n">v</span><span class="p">]]</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="n">v</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">test</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">tfchip</span><span class="p">[</span><span class="n">k</span><span class="p">]))</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="k">if</span> <span class="n">test</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">pre_res</span> <span class="o">=</span> <span class="n">gp</span><span class="o">.</span><span class="n">prerank</span><span class="p">(</span>
                <span class="n">rnk</span><span class="o">=</span><span class="n">test</span><span class="p">,</span>
                <span class="n">gene_sets</span><span class="o">=</span><span class="p">[{</span><span class="n">v</span><span class="p">:</span> <span class="n">tfchip</span><span class="p">[</span><span class="n">k</span><span class="p">]}],</span>
                <span class="n">background</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">grn</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
                <span class="n">min_size</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                <span class="n">max_size</span><span class="o">=</span><span class="mi">4000</span><span class="p">,</span>
                <span class="n">permutation_num</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">val</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">pre_res</span><span class="o">.</span><span class="n">res2d</span><span class="p">[</span>
                <span class="p">(</span><span class="n">pre_res</span><span class="o">.</span><span class="n">res2d</span><span class="p">[</span><span class="s2">&quot;FDR q-val&quot;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mf">0.05</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">pre_res</span><span class="o">.</span><span class="n">res2d</span><span class="p">[</span><span class="s2">&quot;NES&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span>
            <span class="p">]</span>
            <span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;NES&quot;</span><span class="p">],</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Name&quot;</span><span class="p">])</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">val</span><span class="o">.</span><span class="n">Term</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># print(&quot;found! &quot;, val.Term.tolist()[0])</span>
            <span class="c1"># print(pre_res.res2d[&quot;NES&quot;])</span>
            <span class="c1"># print(pre_res.res2d[&quot;FDR q-val&quot;])</span>
            <span class="c1"># print(&quot;\n&quot;)</span>
            <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">pass</span>
            <span class="c1"># print(&quot;no sig...&quot;)</span>
        <span class="n">res</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">pre_res</span><span class="o">.</span><span class="n">res2d</span>
        <span class="c1"># print(val.Term.tolist()[:2])</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">disable</span><span class="p">(</span><span class="n">previous_level</span><span class="p">)</span>
    <span class="n">j</span> <span class="o">=</span> <span class="n">j</span> <span class="k">if</span> <span class="n">j</span> <span class="o">!=</span> <span class="mi">0</span> <span class="k">else</span> <span class="mi">1</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">doplot</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;found some significant results for &quot;</span><span class="p">,</span> <span class="n">i</span> <span class="o">*</span> <span class="mi">100</span> <span class="o">/</span> <span class="n">j</span><span class="p">,</span> <span class="s2">&quot;% TFs</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;_________________________________________&quot;</span><span class="p">)</span>
    <span class="n">metrics</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;significant_enriched_TFtargets&quot;</span><span class="p">:</span> <span class="n">i</span> <span class="o">*</span> <span class="mi">100</span> <span class="o">/</span> <span class="n">j</span><span class="p">})</span>
    <span class="n">metrics</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">compare_to</span><span class="p">(</span><span class="n">to</span><span class="o">=</span><span class="s2">&quot;omnipath&quot;</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="s2">&quot;human&quot;</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">metrics</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>


<div class="doc doc-object doc-function">


<h3 id="bengrn.base.compute_epr" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compute_epr</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>compute_epr computes the Expected Precision Recall (EPR) metric for the given classifier, test data, and true labels.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>clf</code></b>
                  (<code><span title="sklearn.base.ClassifierMixin">ClassifierMixin</span></code>)
              –
              <div class="doc-md-description">
                <p>The classifier to evaluate.</p>
              </div>
            </li>
            <li>
              <b><code>X_test</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>The test data features.</p>
              </div>
            </li>
            <li>
              <b><code>y_test</code></b>
                  (<code><span title="numpy.ndarray">ndarray</span></code>)
              –
              <div class="doc-md-description">
                <p>The true labels for the test data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>float</code></b>(                  <code>float</code>
)              –
              <div class="doc-md-description">
                <p>The computed Expected Precision Recall (EPR) metric.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>bengrn/base.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">925</span>
<span class="normal">926</span>
<span class="normal">927</span>
<span class="normal">928</span>
<span class="normal">929</span>
<span class="normal">930</span>
<span class="normal">931</span>
<span class="normal">932</span>
<span class="normal">933</span>
<span class="normal">934</span>
<span class="normal">935</span>
<span class="normal">936</span>
<span class="normal">937</span>
<span class="normal">938</span>
<span class="normal">939</span>
<span class="normal">940</span>
<span class="normal">941</span>
<span class="normal">942</span>
<span class="normal">943</span>
<span class="normal">944</span>
<span class="normal">945</span>
<span class="normal">946</span>
<span class="normal">947</span>
<span class="normal">948</span>
<span class="normal">949</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">compute_epr</span><span class="p">(</span><span class="n">clf</span><span class="p">,</span> <span class="n">X_test</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">y_test</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    compute_epr computes the Expected Precision Recall (EPR) metric for the given classifier, test data, and true labels.</span>

<span class="sd">    Args:</span>
<span class="sd">        clf (sklearn.base.ClassifierMixin): The classifier to evaluate.</span>
<span class="sd">        X_test (numpy.ndarray): The test data features.</span>
<span class="sd">        y_test (numpy.ndarray): The true labels for the test data.</span>

<span class="sd">    Returns:</span>
<span class="sd">        float: The computed Expected Precision Recall (EPR) metric.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">prb</span> <span class="o">=</span> <span class="n">clf</span><span class="o">.</span><span class="n">predict_proba</span><span class="p">(</span><span class="n">X_test</span><span class="p">)[:,</span> <span class="mi">1</span><span class="p">]</span>

    <span class="n">K</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">y_test</span><span class="p">)</span>
    <span class="c1"># get only the top-K elems from prb</span>
    <span class="n">pred</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">prb</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
    <span class="n">pred</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">prb</span><span class="p">)[</span><span class="o">-</span><span class="nb">int</span><span class="p">(</span><span class="n">K</span><span class="p">)</span> <span class="p">:]]</span> <span class="o">=</span> <span class="mi">1</span>

    <span class="n">true_positive</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">pred</span><span class="p">[</span><span class="n">y_test</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">false_positive</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">pred</span><span class="p">[</span><span class="n">y_test</span> <span class="o">==</span> <span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">false_negative</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">pred</span><span class="p">[</span><span class="n">y_test</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>
    <span class="n">true_negative</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">pred</span><span class="p">[</span><span class="n">y_test</span> <span class="o">==</span> <span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>
    <span class="n">odds_ratio</span> <span class="o">=</span> <span class="p">(</span><span class="n">true_positive</span> <span class="o">*</span> <span class="n">true_negative</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">false_positive</span> <span class="o">*</span> <span class="n">false_negative</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">odds_ratio</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="bengrn.base.compute_genie3" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compute_genie3</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>This function computes the GENIE3 algorithm on the given data.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>adata</code></b>
                  (<code><span title="anndata.AnnData">AnnData</span></code>)
              –
              <div class="doc-md-description">
                <p>The annotated data matrix of shape n_obs x n_vars. Rows correspond to cells and columns to genes.</p>
              </div>
            </li>
            <li>
              <b><code>nthreads</code></b>
                  (<code>int</code>, default:
                      <code>30</code>
)
              –
              <div class="doc-md-description">
                <p>The number of threads to use for computation. Defaults to 30.</p>
              </div>
            </li>
            <li>
              <b><code>ntrees</code></b>
                  (<code>int</code>, default:
                      <code>100</code>
)
              –
              <div class="doc-md-description">
                <p>The number of trees to use for the Random Forests. Defaults to 100.</p>
              </div>
            </li>
            <li>
              <b><code>**kwargs</code></b>
              –
              <div class="doc-md-description">
                <p>Additional arguments to pass to the GENIE3 function.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>GRNAnnData</code></b>(                  <code><span title="grnndata.GRNAnnData">GRNAnnData</span></code>
)              –
              <div class="doc-md-description">
                <p>The Gene Regulatory Network data computed using the GENIE3 algorithm.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>bengrn/base.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">compute_genie3</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">:</span> <span class="n">AnnData</span><span class="p">,</span> <span class="n">nthreads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">30</span><span class="p">,</span> <span class="n">ntrees</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GRNAnnData</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function computes the GENIE3 algorithm on the given data.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata (AnnData): The annotated data matrix of shape n_obs x n_vars. Rows correspond to cells and columns to genes.</span>
<span class="sd">        nthreads (int, optional): The number of threads to use for computation. Defaults to 30.</span>
<span class="sd">        ntrees (int, optional): The number of trees to use for the Random Forests. Defaults to 100.</span>
<span class="sd">        **kwargs: Additional arguments to pass to the GENIE3 function.</span>

<span class="sd">    Returns:</span>
<span class="sd">        GRNAnnData: The Gene Regulatory Network data computed using the GENIE3 algorithm.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mat</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">todense</span><span class="p">())</span>
    <span class="n">names</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">[</span><span class="n">mat</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">var</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">mat</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>
    <span class="n">mat</span> <span class="o">=</span> <span class="n">mat</span><span class="p">[:,</span> <span class="n">mat</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>
    <span class="n">VIM</span> <span class="o">=</span> <span class="n">GENIE3</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="n">gene_names</span><span class="o">=</span><span class="n">names</span><span class="p">,</span> <span class="n">nthreads</span><span class="o">=</span><span class="n">nthreads</span><span class="p">,</span> <span class="n">ntrees</span><span class="o">=</span><span class="n">ntrees</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="n">grn</span> <span class="o">=</span> <span class="n">GRNAnnData</span><span class="p">(</span><span class="n">grn</span><span class="o">=</span><span class="n">VIM</span><span class="p">,</span> <span class="n">X</span><span class="o">=</span><span class="n">mat</span><span class="p">,</span> <span class="n">var</span><span class="o">=</span><span class="n">var</span><span class="p">,</span> <span class="n">obs</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">)</span>
    <span class="n">grn</span><span class="o">.</span><span class="n">var_names</span> <span class="o">=</span> <span class="n">grn</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]</span>
    <span class="n">grn</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;TFs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="kc">True</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">utils</span><span class="o">.</span><span class="n">TF</span> <span class="k">else</span> <span class="kc">False</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">grn</span><span class="o">.</span><span class="n">var_names</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">grn</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="bengrn.base.compute_pr" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compute_pr</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>compute_pr computes the precision and recall metrics for the given GRN and true matrix.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>grn</code></b>
                  (<code><span title="numpy.array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>The Gene Regulatory Network matrix, where each element represents the strength of the regulatory relationship between genes.</p>
              </div>
            </li>
            <li>
              <b><code>true</code></b>
                  (<code><span title="numpy.array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>The ground truth matrix, where each element indicates the presence (1) or absence (0) of a regulatory relationship.</p>
              </div>
            </li>
            <li>
              <b><code>do_auc</code></b>
                  (<code>bool</code>, default:
                      <code>True</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to compute the Area Under the Precision-Recall Curve (AUPRC). Defaults to True.</p>
              </div>
            </li>
            <li>
              <b><code>doplot</code></b>
                  (<code>bool</code>, default:
                      <code>True</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to plot the precision and recall metrics. Defaults to True.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code>ValueError</code>
              –
              <div class="doc-md-description">
                <p>If the shape of the GRN and the true matrix do not match.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>dict</code></b>              –
              <div class="doc-md-description">
                <p>A dictionary containing precision, recall, and random precision metrics.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>bengrn/base.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span>
<span class="normal">804</span>
<span class="normal">805</span>
<span class="normal">806</span>
<span class="normal">807</span>
<span class="normal">808</span>
<span class="normal">809</span>
<span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span>
<span class="normal">820</span>
<span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span>
<span class="normal">826</span>
<span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span>
<span class="normal">830</span>
<span class="normal">831</span>
<span class="normal">832</span>
<span class="normal">833</span>
<span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span>
<span class="normal">838</span>
<span class="normal">839</span>
<span class="normal">840</span>
<span class="normal">841</span>
<span class="normal">842</span>
<span class="normal">843</span>
<span class="normal">844</span>
<span class="normal">845</span>
<span class="normal">846</span>
<span class="normal">847</span>
<span class="normal">848</span>
<span class="normal">849</span>
<span class="normal">850</span>
<span class="normal">851</span>
<span class="normal">852</span>
<span class="normal">853</span>
<span class="normal">854</span>
<span class="normal">855</span>
<span class="normal">856</span>
<span class="normal">857</span>
<span class="normal">858</span>
<span class="normal">859</span>
<span class="normal">860</span>
<span class="normal">861</span>
<span class="normal">862</span>
<span class="normal">863</span>
<span class="normal">864</span>
<span class="normal">865</span>
<span class="normal">866</span>
<span class="normal">867</span>
<span class="normal">868</span>
<span class="normal">869</span>
<span class="normal">870</span>
<span class="normal">871</span>
<span class="normal">872</span>
<span class="normal">873</span>
<span class="normal">874</span>
<span class="normal">875</span>
<span class="normal">876</span>
<span class="normal">877</span>
<span class="normal">878</span>
<span class="normal">879</span>
<span class="normal">880</span>
<span class="normal">881</span>
<span class="normal">882</span>
<span class="normal">883</span>
<span class="normal">884</span>
<span class="normal">885</span>
<span class="normal">886</span>
<span class="normal">887</span>
<span class="normal">888</span>
<span class="normal">889</span>
<span class="normal">890</span>
<span class="normal">891</span>
<span class="normal">892</span>
<span class="normal">893</span>
<span class="normal">894</span>
<span class="normal">895</span>
<span class="normal">896</span>
<span class="normal">897</span>
<span class="normal">898</span>
<span class="normal">899</span>
<span class="normal">900</span>
<span class="normal">901</span>
<span class="normal">902</span>
<span class="normal">903</span>
<span class="normal">904</span>
<span class="normal">905</span>
<span class="normal">906</span>
<span class="normal">907</span>
<span class="normal">908</span>
<span class="normal">909</span>
<span class="normal">910</span>
<span class="normal">911</span>
<span class="normal">912</span>
<span class="normal">913</span>
<span class="normal">914</span>
<span class="normal">915</span>
<span class="normal">916</span>
<span class="normal">917</span>
<span class="normal">918</span>
<span class="normal">919</span>
<span class="normal">920</span>
<span class="normal">921</span>
<span class="normal">922</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">compute_pr</span><span class="p">(</span>
    <span class="n">grn</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">,</span>
    <span class="n">true</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">,</span>
    <span class="n">do_auc</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">doplot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    compute_pr computes the precision and recall metrics for the given GRN and true matrix.</span>

<span class="sd">    Args:</span>
<span class="sd">        grn (np.array): The Gene Regulatory Network matrix, where each element represents the strength of the regulatory relationship between genes.</span>
<span class="sd">        true (np.array): The ground truth matrix, where each element indicates the presence (1) or absence (0) of a regulatory relationship.</span>
<span class="sd">        do_auc (bool, optional): Whether to compute the Area Under the Precision-Recall Curve (AUPRC). Defaults to True.</span>
<span class="sd">        doplot (bool, optional): Whether to plot the precision and recall metrics. Defaults to True.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If the shape of the GRN and the true matrix do not match.</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: A dictionary containing precision, recall, and random precision metrics.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">grn</span><span class="o">.</span><span class="n">shape</span> <span class="o">!=</span> <span class="n">true</span><span class="o">.</span><span class="n">shape</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;The shape of the GRN and the true matrix do not match.&quot;</span><span class="p">)</span>
    <span class="n">metrics</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">true</span> <span class="o">=</span> <span class="n">true</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
    <span class="n">tot</span> <span class="o">=</span> <span class="p">(</span><span class="n">grn</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">grn</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">-</span> <span class="n">grn</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">precision</span> <span class="o">=</span> <span class="p">(</span><span class="n">grn</span><span class="p">[</span><span class="n">true</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">/</span> <span class="p">(</span><span class="n">grn</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="n">recall</span> <span class="o">=</span> <span class="p">(</span><span class="n">grn</span><span class="p">[</span><span class="n">true</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">/</span> <span class="n">true</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="n">rand_prec</span> <span class="o">=</span> <span class="n">true</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">/</span> <span class="n">tot</span>

    <span class="k">if</span> <span class="n">doplot</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span>
            <span class="s2">&quot;precision: &quot;</span><span class="p">,</span>
            <span class="n">precision</span><span class="p">,</span>
            <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">recall: &quot;</span><span class="p">,</span>
            <span class="n">recall</span><span class="p">,</span>
            <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">random precision:&quot;</span><span class="p">,</span>
            <span class="n">rand_prec</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="n">metrics</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s2">&quot;precision&quot;</span><span class="p">:</span> <span class="n">precision</span><span class="p">,</span>
            <span class="s2">&quot;recall&quot;</span><span class="p">:</span> <span class="n">recall</span><span class="p">,</span>
            <span class="s2">&quot;rand_precision&quot;</span><span class="p">:</span> <span class="n">rand_prec</span><span class="p">,</span>
        <span class="p">}</span>
    <span class="p">)</span>
    <span class="c1"># Initialize lists to store precision and recall values</span>
    <span class="n">precision_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">precision</span><span class="p">]</span>
    <span class="n">recall_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">recall</span><span class="p">]</span>
    <span class="c1"># Define the thresholds to vary</span>
    <span class="n">thresholds</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
        <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">101</span><span class="p">)[:</span><span class="o">-</span><span class="mi">2</span><span class="p">],</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">logspace</span><span class="p">(</span><span class="mf">0.99</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">30</span><span class="p">))</span>
    <span class="p">)</span>
    <span class="n">thresholds</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">quantile</span><span class="p">(</span><span class="n">grn</span><span class="p">,</span> <span class="n">thresholds</span><span class="p">)</span>
    <span class="c1"># Calculate precision and recall for each threshold</span>
    <span class="k">if</span> <span class="n">do_auc</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">threshold</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="o">.</span><span class="n">tqdm</span><span class="p">(</span><span class="n">thresholds</span><span class="p">[</span><span class="mi">1</span><span class="p">:]):</span>
            <span class="n">precision</span> <span class="o">=</span> <span class="p">(</span><span class="n">grn</span><span class="p">[</span><span class="n">true</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">threshold</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">/</span> <span class="p">(</span><span class="n">grn</span> <span class="o">&gt;</span> <span class="n">threshold</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
            <span class="n">recall</span> <span class="o">=</span> <span class="p">(</span><span class="n">grn</span><span class="p">[</span><span class="n">true</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">threshold</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">/</span> <span class="n">true</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
            <span class="n">precision_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">precision</span><span class="p">)</span>
            <span class="n">recall_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">recall</span><span class="p">)</span>
        <span class="c1"># Calculate AUPRC by integrating the precision-recall curve</span>
        <span class="k">if</span> <span class="mf">1.0</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">recall_list</span><span class="p">:</span>
            <span class="n">precision_list</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">rand_prec</span><span class="p">)</span>
            <span class="n">recall_list</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">recall_list</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">precision_list</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">rand_prec</span><span class="p">)</span>
            <span class="n">recall_list</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">)</span>
        <span class="n">precision_list</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">precision_list</span><span class="p">))</span>
        <span class="n">recall_list</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">recall_list</span><span class="p">))</span>
        <span class="n">auprc</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">trapz</span><span class="p">(</span><span class="n">precision_list</span><span class="p">,</span> <span class="n">recall_list</span><span class="p">)</span>
        <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;auprc&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">auprc</span>

        <span class="c1"># Compute Average Precision (AP) manually</span>
        <span class="n">sorted_indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="o">-</span><span class="n">grn</span><span class="o">.</span><span class="n">flatten</span><span class="p">())</span>
        <span class="n">sorted_true</span> <span class="o">=</span> <span class="n">true</span><span class="o">.</span><span class="n">flatten</span><span class="p">()[</span><span class="n">sorted_indices</span><span class="p">]</span>

        <span class="n">tp_cumsum</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="n">sorted_true</span><span class="p">)</span>
        <span class="n">fp_cumsum</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">cumsum</span><span class="p">(</span><span class="o">~</span><span class="n">sorted_true</span><span class="p">)</span>

        <span class="n">precision_at_k</span> <span class="o">=</span> <span class="n">tp_cumsum</span> <span class="o">/</span> <span class="p">(</span><span class="n">tp_cumsum</span> <span class="o">+</span> <span class="n">fp_cumsum</span><span class="p">)</span>
        <span class="n">recall_at_k</span> <span class="o">=</span> <span class="n">tp_cumsum</span> <span class="o">/</span> <span class="n">true</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>

        <span class="n">ap</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">precision_at_k</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">diff</span><span class="p">(</span><span class="n">recall_at_k</span><span class="p">))</span>
        <span class="n">metrics</span><span class="p">[</span><span class="s2">&quot;ap&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ap</span>
        <span class="k">if</span> <span class="n">doplot</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Average Precision (AP): &quot;</span><span class="p">,</span> <span class="n">ap</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">doplot</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Area Under Precision-Recall Curve (AUPRC): &quot;</span><span class="p">,</span> <span class="n">auprc</span><span class="p">)</span>

    <span class="c1"># compute EPR</span>
    <span class="c1"># get the indices of the topK highest values in &quot;grn&quot;</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">grn</span><span class="p">,</span> <span class="n">csr_matrix</span><span class="p">):</span>
        <span class="n">grn</span> <span class="o">=</span> <span class="n">grn</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">grn</span><span class="p">,</span> <span class="n">csc_matrix</span><span class="p">):</span>
        <span class="n">grn</span> <span class="o">=</span> <span class="n">grn</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span>
    <span class="n">indices</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argpartition</span><span class="p">(</span><span class="n">grn</span><span class="o">.</span><span class="n">flatten</span><span class="p">(),</span> <span class="o">-</span><span class="nb">int</span><span class="p">(</span><span class="n">true</span><span class="o">.</span><span class="n">sum</span><span class="p">()))[</span><span class="o">-</span><span class="nb">int</span><span class="p">(</span><span class="n">true</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span> <span class="p">:]</span>
    <span class="c1"># Compute the odds ratio</span>
    <span class="n">true_positive</span> <span class="o">=</span> <span class="n">true</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">unravel_index</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">true</span><span class="o">.</span><span class="n">shape</span><span class="p">)]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
    <span class="n">false_positive</span> <span class="o">=</span> <span class="n">true</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">-</span> <span class="n">true_positive</span>
    <span class="c1"># this is normal as we compute on the same number of pred_pos as true_pos</span>
    <span class="n">false_negative</span> <span class="o">=</span> <span class="n">true</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">-</span> <span class="n">true_positive</span>
    <span class="n">true_negative</span> <span class="o">=</span> <span class="n">tot</span> <span class="o">-</span> <span class="n">true_positive</span> <span class="o">-</span> <span class="n">false_positive</span> <span class="o">-</span> <span class="n">false_negative</span>
    <span class="c1"># Avoid division by zero</span>
    <span class="c1"># this is a debugger line</span>
    <span class="k">if</span> <span class="n">true_negative</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">false_positive</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">odds_ratio</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="s2">&quot;inf&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">odds_ratio</span> <span class="o">=</span> <span class="p">(</span><span class="n">true_positive</span> <span class="o">*</span> <span class="n">true_negative</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">false_positive</span> <span class="o">*</span> <span class="n">false_negative</span><span class="p">)</span>

    <span class="n">metrics</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;epr&quot;</span><span class="p">:</span> <span class="n">odds_ratio</span><span class="p">})</span>
    <span class="k">if</span> <span class="n">doplot</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;EPR:&quot;</span><span class="p">,</span> <span class="n">odds_ratio</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span>
            <span class="n">recall_list</span><span class="p">,</span>
            <span class="n">precision_list</span><span class="p">,</span>
            <span class="n">marker</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">,</span>
            <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">,</span>
            <span class="n">color</span><span class="o">=</span><span class="s2">&quot;b&quot;</span><span class="p">,</span>
            <span class="n">label</span><span class="o">=</span><span class="s2">&quot;p-r&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span>
            <span class="p">[</span><span class="n">recall_list</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">recall_list</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]],</span>
            <span class="p">[</span><span class="n">rand_prec</span><span class="p">,</span> <span class="n">rand_prec</span><span class="p">],</span>
            <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span>
            <span class="n">color</span><span class="o">=</span><span class="s2">&quot;r&quot;</span><span class="p">,</span>
            <span class="n">label</span><span class="o">=</span><span class="s2">&quot;Random Precision&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="s2">&quot;lower left&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Precision-Recall Curve&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Recall&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Precision&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">xscale</span><span class="p">(</span><span class="s2">&quot;log&quot;</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">metrics</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="bengrn.base.compute_scenic" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">compute_scenic</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>This function computes the SCENIC algorithm on the given dataset.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>adata</code></b>
                  (<code><span title="anndata.AnnData">AnnData</span></code>)
              –
              <div class="doc-md-description">
                <p>The annotated data matrix of shape n_obs x n_vars. Rows correspond to cells and columns to genes.</p>
              </div>
            </li>
            <li>
              <b><code>data_dir</code></b>
                  (<code>str</code>, default:
                      <code><span title="bengrn.base.FILEDIR">FILEDIR</span> + &#39;/../data&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The directory where the data files will be stored. Defaults to FILEDIR + "/../data".</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>GRNAnnData</code></b>              –
              <div class="doc-md-description">
                <p>The Gene Regulatory Network data.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>bengrn/base.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">compute_scenic</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">data_dir</span><span class="o">=</span><span class="n">FILEDIR</span> <span class="o">+</span> <span class="s2">&quot;/../data&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function computes the SCENIC algorithm on the given dataset.</span>

<span class="sd">    Args:</span>
<span class="sd">        adata (AnnData): The annotated data matrix of shape n_obs x n_vars. Rows correspond to cells and columns to genes.</span>
<span class="sd">        data_dir (str, optional): The directory where the data files will be stored. Defaults to FILEDIR + &quot;/../data&quot;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        GRNAnnData: The Gene Regulatory Network data.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">url1</span> <span class="o">=</span> <span class="s2">&quot;https://resources.aertslab.org/cistarget/motif2tf/motifs-v10nr_clust-nr.hgnc-m0.001-o0.0.tbl&quot;</span>
    <span class="n">url2</span> <span class="o">=</span> <span class="s2">&quot;https://resources.aertslab.org/cistarget/databases/homo_sapiens/hg38/refseq_r80/mc_v10_clust/gene_based/hg38_10kbp_up_10kbp_down_full_tx_v10_clust.genes_vs_motifs.rankings.feather&quot;</span>

    <span class="n">file1</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">url1</span><span class="p">))</span>
    <span class="n">file2</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">data_dir</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">url2</span><span class="p">))</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">file1</span><span class="p">):</span>
        <span class="n">urllib</span><span class="o">.</span><span class="n">request</span><span class="o">.</span><span class="n">urlretrieve</span><span class="p">(</span><span class="n">url1</span><span class="p">,</span> <span class="n">file1</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">file2</span><span class="p">):</span>
        <span class="n">urllib</span><span class="o">.</span><span class="n">request</span><span class="o">.</span><span class="n">urlretrieve</span><span class="p">(</span><span class="n">url2</span><span class="p">,</span> <span class="n">file2</span><span class="p">)</span>
    <span class="n">dbs</span> <span class="o">=</span> <span class="n">RankingDatabase</span><span class="p">(</span><span class="n">fname</span><span class="o">=</span><span class="n">file2</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;genes_10kb_human_rank_v10&quot;</span><span class="p">)</span>
    <span class="n">adjacencies</span> <span class="o">=</span> <span class="n">grnboost2</span><span class="p">(</span>
        <span class="n">expression_data</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">to_df</span><span class="p">(),</span> <span class="n">tf_names</span><span class="o">=</span><span class="n">utils</span><span class="o">.</span><span class="n">TF</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span>
    <span class="p">)</span>
    <span class="n">modules</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">modules_from_adjacencies</span><span class="p">(</span><span class="n">adjacencies</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">to_df</span><span class="p">()))</span>

    <span class="n">df</span> <span class="o">=</span> <span class="n">prune2df</span><span class="p">([</span><span class="n">dbs</span><span class="p">],</span> <span class="n">modules</span><span class="p">,</span> <span class="n">file1</span><span class="p">)</span>
    <span class="n">regulons</span> <span class="o">=</span> <span class="n">df2regulons</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
    <span class="c1"># auc_mtx = aucell(adata.to_df(), regulons, num_workers=8)</span>
    <span class="c1"># sns.clustermap(auc_mtx, figsize=(12,12))</span>

    <span class="c1"># compute the grn matrix</span>
    <span class="n">var_names</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">da</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">var_names</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">var_names</span><span class="p">)),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">reg</span> <span class="ow">in</span> <span class="n">regulons</span><span class="p">:</span>
        <span class="n">i</span> <span class="o">=</span> <span class="n">var_names</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">reg</span><span class="o">.</span><span class="n">transcription_factor</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">reg</span><span class="o">.</span><span class="n">gene2weight</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">da</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">var_names</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">k</span><span class="p">)]</span> <span class="o">=</span> <span class="n">v</span>
    <span class="c1"># create the grndata</span>
    <span class="n">grn</span> <span class="o">=</span> <span class="n">GRNAnnData</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">grn</span><span class="o">=</span><span class="n">da</span><span class="p">)</span>

    <span class="n">da</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">var_names</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">var_names</span><span class="p">)),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">reg</span> <span class="ow">in</span> <span class="n">modules</span><span class="p">:</span>
        <span class="n">i</span> <span class="o">=</span> <span class="n">var_names</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">reg</span><span class="o">.</span><span class="n">transcription_factor</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">reg</span><span class="o">.</span><span class="n">gene2weight</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">da</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">var_names</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">k</span><span class="p">)]</span> <span class="o">=</span> <span class="n">v</span>
    <span class="n">grn</span><span class="o">.</span><span class="n">varp</span><span class="p">[</span><span class="s2">&quot;modules&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">da</span>
    <span class="n">grn</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;TFs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="kc">True</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">utils</span><span class="o">.</span><span class="n">TF</span> <span class="k">else</span> <span class="kc">False</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">grn</span><span class="o">.</span><span class="n">var_names</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">grn</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="bengrn.base.get_GT_db" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_GT_db</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>use_prior_network loads a prior GRN from a list of available networks.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>name</code></b>
                  (<code>str</code>, default:
                      <code>&#39;collectri&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>name of the network to load. Defaults to "collectri".</p>
              </div>
            </li>
            <li>
              <b><code>organism</code></b>
                  (<code>str</code>, default:
                      <code>&#39;human&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>organism to load the network for. Defaults to "human".</p>
              </div>
            </li>
            <li>
              <b><code>split_complexes</code></b>
                  (<code>bool</code>, default:
                      <code>True</code>
)
              –
              <div class="doc-md-description">
                <p>whether to split complexes into individual genes. Defaults to True.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
              –
              <div class="doc-md-description">
                <p>pd.DataFrame: The prior GRN as a pandas DataFrame.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code>ValueError</code>
              –
              <div class="doc-md-description">
                <p>if the provided name is not amongst the available names.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>bengrn/base.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_GT_db</span><span class="p">(</span>
    <span class="n">name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;collectri&quot;</span><span class="p">,</span> <span class="n">organism</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;human&quot;</span><span class="p">,</span> <span class="n">split_complexes</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    use_prior_network loads a prior GRN from a list of available networks.</span>

<span class="sd">    Args:</span>
<span class="sd">        name (str, optional): name of the network to load. Defaults to &quot;collectri&quot;.</span>
<span class="sd">        organism (str, optional): organism to load the network for. Defaults to &quot;human&quot;.</span>
<span class="sd">        split_complexes (bool, optional): whether to split complexes into individual genes. Defaults to True.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: The prior GRN as a pandas DataFrame.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: if the provided name is not amongst the available names.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;tflink&quot;</span><span class="p">:</span>
        <span class="n">TFLINK</span> <span class="o">=</span> <span class="s2">&quot;https://cdn.netbiol.org/tflink/download_files/TFLink_Homo_sapiens_interactions_All_simpleFormat_v1.0.tsv.gz&quot;</span>
        <span class="n">net</span> <span class="o">=</span> <span class="n">pd_load_cached</span><span class="p">(</span><span class="n">TFLINK</span><span class="p">)</span>
        <span class="n">net</span> <span class="o">=</span> <span class="n">net</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Name.TF&quot;</span><span class="p">:</span> <span class="s2">&quot;source&quot;</span><span class="p">,</span> <span class="s2">&quot;Name.Target&quot;</span><span class="p">:</span> <span class="s2">&quot;target&quot;</span><span class="p">})</span>
    <span class="k">elif</span> <span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;htftarget&quot;</span><span class="p">:</span>
        <span class="n">HTFTARGET</span> <span class="o">=</span> <span class="s2">&quot;http://bioinfo.life.hust.edu.cn/static/hTFtarget/file_download/tf-target-infomation.txt&quot;</span>
        <span class="n">net</span> <span class="o">=</span> <span class="n">pd_load_cached</span><span class="p">(</span><span class="n">HTFTARGET</span><span class="p">)</span>
        <span class="n">net</span> <span class="o">=</span> <span class="n">net</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;TF&quot;</span><span class="p">:</span> <span class="s2">&quot;source&quot;</span><span class="p">})</span>
    <span class="k">elif</span> <span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;collectri&quot;</span><span class="p">:</span>
        <span class="kn">import</span> <span class="nn">decoupler</span> <span class="k">as</span> <span class="nn">dc</span>

        <span class="n">net</span> <span class="o">=</span> <span class="n">dc</span><span class="o">.</span><span class="n">get_collectri</span><span class="p">(</span><span class="n">organism</span><span class="o">=</span><span class="n">organism</span><span class="p">,</span> <span class="n">split_complexes</span><span class="o">=</span><span class="n">split_complexes</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span>
            <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;PMID&quot;</span><span class="p">]</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;dorothea&quot;</span><span class="p">:</span>
        <span class="kn">import</span> <span class="nn">decoupler</span> <span class="k">as</span> <span class="nn">dc</span>

        <span class="n">net</span> <span class="o">=</span> <span class="n">dc</span><span class="o">.</span><span class="n">get_dorothea</span><span class="p">(</span><span class="n">organism</span><span class="o">=</span><span class="n">organism</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;omnipath&quot;</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">FILEDIR</span> <span class="o">+</span> <span class="s2">&quot;/../data/omnipath.parquet&quot;</span><span class="p">):</span>
            <span class="kn">from</span> <span class="nn">omnipath.interactions</span> <span class="kn">import</span> <span class="n">AllInteractions</span>
            <span class="kn">from</span> <span class="nn">omnipath.requests</span> <span class="kn">import</span> <span class="n">Annotations</span>

            <span class="n">interactions</span> <span class="o">=</span> <span class="n">AllInteractions</span><span class="p">()</span>
            <span class="n">net</span> <span class="o">=</span> <span class="n">interactions</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">exclude</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;small_molecule&quot;</span><span class="p">,</span> <span class="s2">&quot;lncrna_mrna&quot;</span><span class="p">])</span>
            <span class="n">hgnc</span> <span class="o">=</span> <span class="n">Annotations</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">resources</span><span class="o">=</span><span class="s2">&quot;HGNC&quot;</span><span class="p">)</span>
            <span class="n">rename</span> <span class="o">=</span> <span class="p">{</span><span class="n">v</span><span class="o">.</span><span class="n">uniprot</span><span class="p">:</span> <span class="n">v</span><span class="o">.</span><span class="n">genesymbol</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">hgnc</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()}</span>
            <span class="n">net</span><span class="o">.</span><span class="n">source</span> <span class="o">=</span> <span class="n">net</span><span class="o">.</span><span class="n">source</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">rename</span><span class="p">)</span>
            <span class="n">net</span><span class="o">.</span><span class="n">target</span> <span class="o">=</span> <span class="n">net</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">rename</span><span class="p">)</span>
            <span class="n">net</span><span class="o">.</span><span class="n">to_parquet</span><span class="p">(</span><span class="n">FILEDIR</span> <span class="o">+</span> <span class="s2">&quot;/../data/omnipath.parquet&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">net</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_parquet</span><span class="p">(</span><span class="n">FILEDIR</span> <span class="o">+</span> <span class="s2">&quot;/../data/omnipath.parquet&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;provided name: &#39;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">&#39; is not amongst the available names.&quot;</span><span class="p">)</span>
    <span class="c1"># varnames = list(set(net.iloc[:, :2].values.flatten()))</span>
    <span class="c1"># adata = AnnData(var=varnames)</span>
    <span class="c1"># adata.var_names = varnames</span>
    <span class="c1"># grn = from_adata_and_longform(adata, net, has_weight=True)</span>
    <span class="c1"># return grn</span>
    <span class="k">return</span> <span class="n">net</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="bengrn.base.get_perturb_gt" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_perturb_gt</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>get_perturb_gt retrieves the genome wide perturb seq ground truth data.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>url_bh</code></b>
                  (<code>str</code>, default:
                      <code>&#39;https://plus.figshare.com/ndownloader/files/38349308&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The URL to download the BH-corrected data. Defaults to "https://plus.figshare.com/ndownloader/files/38349308".</p>
              </div>
            </li>
            <li>
              <b><code>filename_bh</code></b>
                  (<code>str</code>, default:
                      <code><span title="bengrn.base.FILEDIR">FILEDIR</span> + &#39;/../data/BH-corrected.csv.gz&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The local filename to save the BH-corrected data. Defaults to FILEDIR + "/../data/BH-corrected.csv.gz".</p>
              </div>
            </li>
            <li>
              <b><code>url_adata</code></b>
                  (<code>str</code>, default:
                      <code>&#39;https://plus.figshare.com/ndownloader/files/35773219&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The URL to download the single-cell perturbation data. Defaults to "https://plus.figshare.com/ndownloader/files/35773219".</p>
              </div>
            </li>
            <li>
              <b><code>filename_adata</code></b>
                  (<code>str</code>, default:
                      <code><span title="bengrn.base.FILEDIR">FILEDIR</span> + &#39;/../data/ess_perturb_sc.h5ad&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The local filename to save the single-cell perturbation data. Defaults to FILEDIR + "/../data/ess_perturb_sc.h5ad".</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>GRNAnnData</code></b>              –
              <div class="doc-md-description">
                <p>The Gene Regulatory Network data as a GRNAnnData object.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>bengrn/base.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_perturb_gt</span><span class="p">(</span>
    <span class="n">url_bh</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;https://plus.figshare.com/ndownloader/files/38349308&quot;</span><span class="p">,</span>
    <span class="n">filename_bh</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">FILEDIR</span> <span class="o">+</span> <span class="s2">&quot;/../data/BH-corrected.csv.gz&quot;</span><span class="p">,</span>
    <span class="n">url_adata</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;https://plus.figshare.com/ndownloader/files/35773219&quot;</span><span class="p">,</span>
    <span class="n">filename_adata</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">FILEDIR</span> <span class="o">+</span> <span class="s2">&quot;/../data/ess_perturb_sc.h5ad&quot;</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    get_perturb_gt retrieves the genome wide perturb seq ground truth data.</span>

<span class="sd">    Args:</span>
<span class="sd">        url_bh (str, optional): The URL to download the BH-corrected data. Defaults to &quot;https://plus.figshare.com/ndownloader/files/38349308&quot;.</span>
<span class="sd">        filename_bh (str, optional): The local filename to save the BH-corrected data. Defaults to FILEDIR + &quot;/../data/BH-corrected.csv.gz&quot;.</span>
<span class="sd">        url_adata (str, optional): The URL to download the single-cell perturbation data. Defaults to &quot;https://plus.figshare.com/ndownloader/files/35773219&quot;.</span>
<span class="sd">        filename_adata (str, optional): The local filename to save the single-cell perturbation data. Defaults to FILEDIR + &quot;/../data/ess_perturb_sc.h5ad&quot;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        GRNAnnData: The Gene Regulatory Network data as a GRNAnnData object.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">filename_bh</span><span class="p">):</span>
        <span class="n">urllib</span><span class="o">.</span><span class="n">request</span><span class="o">.</span><span class="n">urlretrieve</span><span class="p">(</span><span class="n">url_bh</span><span class="p">,</span> <span class="n">filename_bh</span><span class="p">)</span>
    <span class="n">pert</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">filename_bh</span><span class="p">)</span>
    <span class="n">pert</span> <span class="o">=</span> <span class="n">pert</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;Unnamed: 0&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
    <span class="n">pert</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pert</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
    <span class="n">pert</span> <span class="o">=</span> <span class="n">pert</span><span class="p">[</span><span class="o">~</span><span class="n">pert</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">duplicated</span><span class="p">(</span><span class="n">keep</span><span class="o">=</span><span class="s2">&quot;first&quot;</span><span class="p">)]</span><span class="o">.</span><span class="n">T</span>
    <span class="n">pert</span> <span class="o">=</span> <span class="n">pert</span> <span class="o">&lt;</span> <span class="mf">0.05</span>
    <span class="n">adata_sc</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span>
        <span class="n">filename_adata</span><span class="p">,</span>
        <span class="n">backup_url</span><span class="o">=</span><span class="n">url_adata</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">adata_sc</span> <span class="o">=</span> <span class="n">adata_sc</span><span class="p">[</span><span class="n">adata_sc</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">gene_id</span> <span class="o">==</span> <span class="s2">&quot;non-targeting&quot;</span><span class="p">]</span>
    <span class="n">adata_sc</span><span class="p">[:,</span> <span class="n">adata_sc</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">pert</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="o">|</span> <span class="nb">set</span><span class="p">(</span><span class="n">pert</span><span class="o">.</span><span class="n">columns</span><span class="p">))]</span>
    <span class="n">adata_sc</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;organism_ontology_term_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;NCBITaxon:9606&quot;</span>
    <span class="n">adata_sc</span> <span class="o">=</span> <span class="n">adata_sc</span><span class="p">[:,</span> <span class="n">adata_sc</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>

    <span class="n">pert</span> <span class="o">=</span> <span class="n">pert</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pert</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">adata_sc</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">)]</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
        <span class="p">:,</span> <span class="n">pert</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">adata_sc</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
    <span class="p">]</span>

    <span class="n">missing_indices</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">adata_sc</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">pert</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
    <span class="n">pert</span> <span class="o">=</span> <span class="n">pert</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">pert</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">missing_indices</span><span class="p">),</span> <span class="n">fill_value</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">missing_indices</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">adata_sc</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">pert</span><span class="o">.</span><span class="n">columns</span><span class="p">))</span>
    <span class="n">pert</span> <span class="o">=</span> <span class="n">pert</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">pert</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">missing_indices</span><span class="p">),</span> <span class="n">fill_value</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="n">pert</span> <span class="o">=</span> <span class="n">pert</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">adata_sc</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">]</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">adata_sc</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">GRNAnnData</span><span class="p">(</span>
        <span class="n">X</span><span class="o">=</span><span class="n">csr_matrix</span><span class="p">(</span><span class="n">adata_sc</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">toarray</span><span class="p">()),</span>
        <span class="n">var</span><span class="o">=</span><span class="n">adata_sc</span><span class="o">.</span><span class="n">var</span><span class="p">,</span>
        <span class="n">obs</span><span class="o">=</span><span class="n">adata_sc</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span>
        <span class="n">grn</span><span class="o">=</span><span class="n">csr_matrix</span><span class="p">(</span><span class="n">pert</span><span class="o">.</span><span class="n">values</span><span class="p">),</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="bengrn.base.get_scenicplus" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_scenicplus</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>This function retrieves a loomx scenicplus data from a given file path and loads it as a GrnnData</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>filepath</code></b>
              –
              <div class="doc-md-description">
                <p>str, optional
The path to the scenicplus data file.
Default is FILEDIR + "/../data/10xPBMC_homo_scenicplus_genebased_scope.loom".</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Raises:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code>FileNotFoundError</code>
              –
              <div class="doc-md-description">
                <p>If the file at the given path does not exist.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>GrnAnnData</code></b>              –
              <div class="doc-md-description">
                <p>The scenicplus data from the given file as a grnndata object</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>bengrn/base.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_scenicplus</span><span class="p">(</span>
    <span class="n">filepath</span><span class="o">=</span><span class="n">FILEDIR</span> <span class="o">+</span> <span class="s2">&quot;/../data/10xPBMC_homo_scenicplus_genebased_scope.loom&quot;</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function retrieves a loomx scenicplus data from a given file path and loads it as a GrnnData</span>

<span class="sd">    Args:</span>
<span class="sd">        filepath : str, optional</span>
<span class="sd">            The path to the scenicplus data file.</span>
<span class="sd">            Default is FILEDIR + &quot;/../data/10xPBMC_homo_scenicplus_genebased_scope.loom&quot;.</span>

<span class="sd">    Raises:</span>
<span class="sd">        FileNotFoundError: If the file at the given path does not exist.</span>

<span class="sd">    Returns:</span>
<span class="sd">        GrnAnnData: The scenicplus data from the given file as a grnndata object</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">filepath</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;The file </span><span class="si">{</span><span class="n">filepath</span><span class="si">}</span><span class="s2"> does not exist. You likely need to download </span><span class="se">\</span>
<span class="s2">                this or another loomxfile from the scope website&quot;</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="n">from_scope_loomfile</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="bengrn.base.get_sroy_gt" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_sroy_gt</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>This function retrieves the ground truth data from the McCall et al.'s paper.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>get</code></b>
                  (<code>str</code>, default:
                      <code>&#39;main&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The specific dataset to retrieve. Options include "main", "liu", and "chen".</p>
              </div>
            </li>
            <li>
              <b><code>join</code></b>
                  (<code>str</code>, default:
                      <code>&#39;outer&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The type of join to be performed when concatenating the data. Default is "outer".</p>
              </div>
            </li>
            <li>
              <b><code>species</code></b>
                  (<code>str</code>, default:
                      <code>&#39;human&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The species of the dataset. Default is "human".</p>
              </div>
            </li>
            <li>
              <b><code>gt</code></b>
                  (<code>str</code>, default:
                      <code>&#39;full&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The type of ground truth data to retrieve. Options include "full", "chip", and "ko". Default is "full".</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>GrnAnnData</code></b>(                  <code>GrnAnnData</code>
)              –
              <div class="doc-md-description">
                <p>The ground truth data as a grnndata object</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>bengrn/base.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_sroy_gt</span><span class="p">(</span>
    <span class="n">get</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;main&quot;</span><span class="p">,</span> <span class="n">join</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;outer&quot;</span><span class="p">,</span> <span class="n">species</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;human&quot;</span><span class="p">,</span> <span class="n">gt</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;full&quot;</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">GrnAnnData</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function retrieves the ground truth data from the McCall et al.&#39;s paper.</span>

<span class="sd">    Args:</span>
<span class="sd">        get (str): The specific dataset to retrieve. Options include &quot;main&quot;, &quot;liu&quot;, and &quot;chen&quot;.</span>
<span class="sd">        join (str, optional): The type of join to be performed when concatenating the data. Default is &quot;outer&quot;.</span>
<span class="sd">        species (str, optional): The species of the dataset. Default is &quot;human&quot;.</span>
<span class="sd">        gt (str, optional): The type of ground truth data to retrieve. Options include &quot;full&quot;, &quot;chip&quot;, and &quot;ko&quot;. Default is &quot;full&quot;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        GrnAnnData: The ground truth data as a grnndata object</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">species</span> <span class="o">==</span> <span class="s2">&quot;human&quot;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">gt</span> <span class="o">==</span> <span class="s2">&quot;full&quot;</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                <span class="n">FILEDIR</span> <span class="o">+</span> <span class="s2">&quot;/../data/GroundTruth/stone_and_sroy/hESC_ground_truth.tsv&quot;</span><span class="p">,</span>
                <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">elif</span> <span class="n">gt</span> <span class="o">==</span> <span class="s2">&quot;chip&quot;</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                <span class="n">FILEDIR</span>
                <span class="o">+</span> <span class="s2">&quot;/../data/GroundTruth/stone_and_sroy/gold_standards/hESC/hESC_chipunion.txt&quot;</span><span class="p">,</span>
                <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">elif</span> <span class="n">gt</span> <span class="o">==</span> <span class="s2">&quot;ko&quot;</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                <span class="n">FILEDIR</span>
                <span class="o">+</span> <span class="s2">&quot;/../data/GroundTruth/stone_and_sroy/gold_standards/hESC/hESC_KDUnion.txt&quot;</span><span class="p">,</span>
                <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">get</span> <span class="o">==</span> <span class="s2">&quot;liu&quot;</span><span class="p">:</span>
            <span class="n">adata</span> <span class="o">=</span> <span class="n">AnnData</span><span class="p">(</span>
                <span class="p">(</span>
                    <span class="mi">2</span>
                    <span class="o">**</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                        <span class="n">FILEDIR</span>
                        <span class="o">+</span> <span class="s2">&quot;/../data/GroundTruth/stone_and_sroy/scRNA/liu_rna_filtered_log2.tsv.gz&quot;</span><span class="p">,</span>
                        <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="o">-</span> <span class="mi">1</span>
            <span class="p">)</span><span class="o">.</span><span class="n">T</span>
        <span class="k">elif</span> <span class="n">get</span> <span class="o">==</span> <span class="s2">&quot;chen&quot;</span><span class="p">:</span>
            <span class="n">adata</span> <span class="o">=</span> <span class="n">AnnData</span><span class="p">(</span>
                <span class="p">(</span>
                    <span class="mi">2</span>
                    <span class="o">**</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                        <span class="n">FILEDIR</span>
                        <span class="o">+</span> <span class="s2">&quot;/../data/GroundTruth/stone_and_sroy/scRNA/chen_rna_filtered_log2.tsv.gz&quot;</span><span class="p">,</span>
                        <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="o">-</span> <span class="mi">1</span>
            <span class="p">)</span><span class="o">.</span><span class="n">T</span>
        <span class="k">elif</span> <span class="n">get</span> <span class="o">==</span> <span class="s2">&quot;han&quot;</span><span class="p">:</span>
            <span class="n">adata</span> <span class="o">=</span> <span class="n">AnnData</span><span class="p">(</span>
                <span class="n">unnormalize</span><span class="p">(</span>
                    <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                        <span class="n">FILEDIR</span>
                        <span class="o">+</span> <span class="s2">&quot;/../data/GroundTruth/stone_and_sroy/scRNA/human_han_GSE107552.csv.gz&quot;</span><span class="p">,</span>
                    <span class="p">)</span>
                    <span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;Cell&quot;</span><span class="p">)</span>
                    <span class="o">.</span><span class="n">T</span><span class="p">,</span>
                    <span class="n">is_root</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="k">elif</span> <span class="n">get</span> <span class="o">==</span> <span class="s2">&quot;mine&quot;</span><span class="p">:</span>
            <span class="c1"># https://www.ebi.ac.uk/gxa/sc/experiments/E-GEOD-36552/downloads</span>
            <span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_mtx</span><span class="p">(</span>
                <span class="n">FILEDIR</span>
                <span class="o">+</span> <span class="s2">&quot;/../data/GroundTruth/stone_and_sroy/scRNA/E-GEOD-36552.aggregated_filtered_counts.mtx&quot;</span>
            <span class="p">)</span><span class="o">.</span><span class="n">T</span>
            <span class="n">col</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                <span class="n">FILEDIR</span>
                <span class="o">+</span> <span class="s2">&quot;/../data/GroundTruth/stone_and_sroy/scRNA/E-GEOD-36552.aggregated_filtered_counts.mtx_rows&quot;</span><span class="p">,</span>
                <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">col</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">genesdf</span> <span class="o">=</span> <span class="n">load_genes</span><span class="p">()</span>
            <span class="n">intersect_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">)</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">genesdf</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
            <span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="nb">list</span><span class="p">(</span><span class="n">intersect_genes</span><span class="p">)]</span>
            <span class="n">genesdf</span> <span class="o">=</span> <span class="n">genesdf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
            <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;ensembl_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span>
            <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">make_index_unique</span><span class="p">(</span><span class="n">genesdf</span><span class="p">[</span><span class="s2">&quot;symbol&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;get must be one of &#39;liu&#39;, &#39;chen&#39;, &#39;han&#39;, or &#39;mine&#39;&quot;</span><span class="p">)</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;organism_ontology_term_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;NCBITaxon:9606&quot;</span>
    <span class="k">elif</span> <span class="n">species</span> <span class="o">==</span> <span class="s2">&quot;mouse&quot;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">gt</span> <span class="o">==</span> <span class="s2">&quot;full&quot;</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                <span class="n">FILEDIR</span> <span class="o">+</span> <span class="s2">&quot;/../data/GroundTruth/stone_and_sroy/mESC_ground_truth.tsv&quot;</span><span class="p">,</span>
                <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">elif</span> <span class="n">gt</span> <span class="o">==</span> <span class="s2">&quot;chip&quot;</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                <span class="n">FILEDIR</span>
                <span class="o">+</span> <span class="s2">&quot;/../data/GroundTruth/stone_and_sroy/gold_standards/mESC/mESC_chipunion.txt&quot;</span><span class="p">,</span>
                <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">elif</span> <span class="n">gt</span> <span class="o">==</span> <span class="s2">&quot;ko&quot;</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                <span class="n">FILEDIR</span>
                <span class="o">+</span> <span class="s2">&quot;/../data/GroundTruth/stone_and_sroy/gold_standards/mESC/mESC_KDUnion.txt&quot;</span><span class="p">,</span>
                <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">get</span> <span class="o">==</span> <span class="s2">&quot;duren&quot;</span><span class="p">:</span>
            <span class="n">adata</span> <span class="o">=</span> <span class="n">AnnData</span><span class="p">(</span>
                <span class="p">(</span>
                    <span class="mi">2</span>
                    <span class="o">**</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                        <span class="n">FILEDIR</span>
                        <span class="o">+</span> <span class="s2">&quot;/../data/GroundTruth/stone_and_sroy/scRNA/duren_rna_filtered_log2.tsv.gz&quot;</span><span class="p">,</span>
                        <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="o">-</span> <span class="mi">1</span>
            <span class="p">)</span><span class="o">.</span><span class="n">T</span>
        <span class="k">elif</span> <span class="n">get</span> <span class="o">==</span> <span class="s2">&quot;semrau&quot;</span><span class="p">:</span>
            <span class="n">adata</span> <span class="o">=</span> <span class="n">AnnData</span><span class="p">(</span>
                <span class="p">(</span>
                    <span class="mi">2</span>
                    <span class="o">**</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                        <span class="n">FILEDIR</span>
                        <span class="o">+</span> <span class="s2">&quot;/../data/GroundTruth/stone_and_sroy/scRNA/semrau_rna_filtered_log2.tsv.gz&quot;</span><span class="p">,</span>
                        <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="o">-</span> <span class="mi">1</span>
            <span class="p">)</span><span class="o">.</span><span class="n">T</span>
        <span class="k">elif</span> <span class="n">get</span> <span class="o">==</span> <span class="s2">&quot;tran&quot;</span><span class="p">:</span>
            <span class="n">adata</span> <span class="o">=</span> <span class="n">AnnData</span><span class="p">(</span>
                <span class="n">unnormalize</span><span class="p">(</span>
                    <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                        <span class="n">FILEDIR</span>
                        <span class="o">+</span> <span class="s2">&quot;/../data/GroundTruth/stone_and_sroy/scRNA/mouse_tran_A2S.csv.gz&quot;</span><span class="p">,</span>
                    <span class="p">)</span>
                    <span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;Cell&quot;</span><span class="p">)</span>
                    <span class="o">.</span><span class="n">T</span><span class="p">,</span>
                    <span class="n">is_root</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="k">elif</span> <span class="n">get</span> <span class="o">==</span> <span class="s2">&quot;zhao&quot;</span><span class="p">:</span>
            <span class="n">adata</span> <span class="o">=</span> <span class="n">AnnData</span><span class="p">(</span>
                <span class="n">unnormalize</span><span class="p">(</span>
                    <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                        <span class="n">FILEDIR</span>
                        <span class="o">+</span> <span class="s2">&quot;/../data/GroundTruth/stone_and_sroy/scRNA/mouse_zhao_GSE114952.csv.gz&quot;</span><span class="p">,</span>
                    <span class="p">)</span>
                    <span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;Cell&quot;</span><span class="p">)</span>
                    <span class="o">.</span><span class="n">T</span><span class="p">,</span>
                    <span class="n">is_root</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;get must be one of &#39;duren&#39;, &#39;semrau&#39;, &#39;tran&#39;, or &#39;zhao&#39;&quot;</span><span class="p">)</span>
        <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;organism_ontology_term_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;NCBITaxon:10090&quot;</span>
    <span class="k">return</span> <span class="n">from_adata_and_longform</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">df</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="bengrn.base.load_genes" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">load_genes</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>load_genes loads the genes for the given organisms.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>organisms</code></b>
                  (<code><span title="typing.Union">Union</span>[str, list]</code>, default:
                      <code>&#39;NCBITaxon:9606&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The organism(s) to load genes for. Can be a single organism string or a list of organism strings. Defaults to "NCBITaxon:9606".</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              –
              <div class="doc-md-description">
                <p>pd.DataFrame: A DataFrame containing gene information for the specified organisms, including columns for gene symbols, mitochondrial genes, ribosomal genes, hemoglobin genes, and organism.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>bengrn/base.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">960</span>
<span class="normal">961</span>
<span class="normal">962</span>
<span class="normal">963</span>
<span class="normal">964</span>
<span class="normal">965</span>
<span class="normal">966</span>
<span class="normal">967</span>
<span class="normal">968</span>
<span class="normal">969</span>
<span class="normal">970</span>
<span class="normal">971</span>
<span class="normal">972</span>
<span class="normal">973</span>
<span class="normal">974</span>
<span class="normal">975</span>
<span class="normal">976</span>
<span class="normal">977</span>
<span class="normal">978</span>
<span class="normal">979</span>
<span class="normal">980</span>
<span class="normal">981</span>
<span class="normal">982</span>
<span class="normal">983</span>
<span class="normal">984</span>
<span class="normal">985</span>
<span class="normal">986</span>
<span class="normal">987</span>
<span class="normal">988</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">load_genes</span><span class="p">(</span><span class="n">organisms</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">list</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;NCBITaxon:9606&quot;</span><span class="p">):</span>  <span class="c1"># &quot;NCBITaxon:10090&quot;,</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    load_genes loads the genes for the given organisms.</span>

<span class="sd">    Args:</span>
<span class="sd">        organisms (Union[str, list], optional): The organism(s) to load genes for. Can be a single organism string or a list of organism strings. Defaults to &quot;NCBITaxon:9606&quot;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: A DataFrame containing gene information for the specified organisms, including columns for gene symbols, mitochondrial genes, ribosomal genes, hemoglobin genes, and organism.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">organismdf</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">organisms</span><span class="p">)</span> <span class="o">==</span> <span class="nb">str</span><span class="p">:</span>
        <span class="n">organisms</span> <span class="o">=</span> <span class="p">[</span><span class="n">organisms</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">organism</span> <span class="ow">in</span> <span class="n">organisms</span><span class="p">:</span>
        <span class="n">genesdf</span> <span class="o">=</span> <span class="n">bt</span><span class="o">.</span><span class="n">Gene</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">organism_id</span><span class="o">=</span><span class="n">bt</span><span class="o">.</span><span class="n">Organism</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">ontology_id</span><span class="o">=</span><span class="n">organism</span><span class="p">)</span><span class="o">.</span><span class="n">first</span><span class="p">()</span><span class="o">.</span><span class="n">id</span>
        <span class="p">)</span><span class="o">.</span><span class="n">df</span><span class="p">()</span>
        <span class="n">genesdf</span> <span class="o">=</span> <span class="n">genesdf</span><span class="p">[</span><span class="o">~</span><span class="n">genesdf</span><span class="p">[</span><span class="s2">&quot;public_source_id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()]</span>
        <span class="n">genesdf</span> <span class="o">=</span> <span class="n">genesdf</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="s2">&quot;ensembl_gene_id&quot;</span><span class="p">)</span>
        <span class="n">genesdf</span> <span class="o">=</span> <span class="n">genesdf</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;ensembl_gene_id&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">sort_index</span><span class="p">()</span>
        <span class="c1"># mitochondrial genes</span>
        <span class="n">genesdf</span><span class="p">[</span><span class="s2">&quot;mt&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genesdf</span><span class="o">.</span><span class="n">symbol</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;MT-&quot;</span><span class="p">)</span>
        <span class="c1"># ribosomal genes</span>
        <span class="n">genesdf</span><span class="p">[</span><span class="s2">&quot;ribo&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genesdf</span><span class="o">.</span><span class="n">symbol</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">((</span><span class="s2">&quot;RPS&quot;</span><span class="p">,</span> <span class="s2">&quot;RPL&quot;</span><span class="p">))</span>
        <span class="c1"># hemoglobin genes.</span>
        <span class="n">genesdf</span><span class="p">[</span><span class="s2">&quot;hb&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">genesdf</span><span class="o">.</span><span class="n">symbol</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">((</span><span class="s2">&quot;^HB[^(P)]&quot;</span><span class="p">))</span>
        <span class="n">genesdf</span><span class="p">[</span><span class="s2">&quot;organism&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">organism</span>
        <span class="n">organismdf</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">genesdf</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">organismdf</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="bengrn.base.precision_recall" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">precision_recall</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>Calculate precision and recall from the true and predicted connections.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>true_con</code></b>
                  (<code><span title="numpy.array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>The true connections.</p>
              </div>
            </li>
            <li>
              <b><code>grn_con</code></b>
                  (<code><span title="numpy.array">array</span></code>)
              –
              <div class="doc-md-description">
                <p>The predicted connections.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
<b><code>float</code></b>              –
              <div class="doc-md-description">
                <p>The precision value.</p>
              </div>
            </li>
            <li>
<b><code>float</code></b>              –
              <div class="doc-md-description">
                <p>The recall value.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>bengrn/base.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">precision_recall</span><span class="p">(</span><span class="n">true_con</span><span class="p">,</span> <span class="n">grn_con</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate precision and recall from the true and predicted connections.</span>

<span class="sd">    Args:</span>
<span class="sd">        true_con (np.array): The true connections.</span>
<span class="sd">        grn_con (np.array): The predicted connections.</span>

<span class="sd">    Returns:</span>
<span class="sd">        float: The precision value.</span>
<span class="sd">        float: The recall value.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">tp</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">true_con</span> <span class="o">&amp;</span> <span class="n">grn_con</span><span class="p">)</span>
    <span class="n">fp</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">grn_con</span> <span class="o">-</span> <span class="n">true_con</span><span class="p">)</span>
    <span class="n">fn</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">true_con</span> <span class="o">-</span> <span class="n">grn_con</span><span class="p">)</span>

    <span class="n">precision</span> <span class="o">=</span> <span class="n">tp</span> <span class="o">/</span> <span class="p">(</span><span class="n">tp</span> <span class="o">+</span> <span class="n">fp</span><span class="p">)</span>
    <span class="n">recall</span> <span class="o">=</span> <span class="n">tp</span> <span class="o">/</span> <span class="p">(</span><span class="n">tp</span> <span class="o">+</span> <span class="n">fn</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">precision</span><span class="p">,</span> <span class="n">recall</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="bengrn.base.train_classifier" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">train_classifier</span></code>

</h3>


    <div class="doc doc-contents ">

      <p>train_classifier trains a classifier to generate a GRN that maps to the ground truth.</p>
<p>Uses a RidgeClassifier to select the best combination of networks to predict the ground truth.
It is used for the head classification part in the scPRINT paper.</p>


<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Parameters:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
              <b><code>grn</code></b>
                  (<code><span title="grnndata.GRNAnnData">GRNAnnData</span></code>)
              –
              <div class="doc-md-description">
                <p>The Gene Regulatory Network data.</p>
              </div>
            </li>
            <li>
              <b><code>gt</code></b>
                  (<code>str</code>, default:
                      <code>&#39;omnipath&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The name of the ground truth database to use. Defaults to "omnipath".</p>
              </div>
            </li>
            <li>
              <b><code>other</code></b>
                  (<code><span title="grnndata.GRNAnnData">GRNAnnData</span></code>, default:
                      <code>None</code>
)
              –
              <div class="doc-md-description">
                <p>Another GRN to compare against. Defaults to None.</p>
              </div>
            </li>
            <li>
              <b><code>use_col</code></b>
                  (<code>str</code>, default:
                      <code>&#39;symbol&#39;</code>
)
              –
              <div class="doc-md-description">
                <p>The column name to use for gene symbols. Defaults to "symbol".</p>
              </div>
            </li>
            <li>
              <b><code>train_size</code></b>
                  (<code>float</code>, default:
                      <code>0.2</code>
)
              –
              <div class="doc-md-description">
                <p>The proportion of the dataset to include in the train split. Defaults to 0.2.</p>
              </div>
            </li>
            <li>
              <b><code>doplot</code></b>
                  (<code>bool</code>, default:
                      <code>True</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to plot the results. Defaults to True.</p>
              </div>
            </li>
            <li>
              <b><code>class_weight</code></b>
                  (<code>dict</code>, default:
                      <code>{1: 200, 0: 1}</code>
)
              –
              <div class="doc-md-description">
                <p>Weights associated with classes in the form {class_label: weight}. Defaults to {1: 200, 0: 1}.</p>
              </div>
            </li>
            <li>
              <b><code>max_iter</code></b>
                  (<code>int</code>, default:
                      <code>1000</code>
)
              –
              <div class="doc-md-description">
                <p>Maximum number of iterations for the classifier. Defaults to 1_000.</p>
              </div>
            </li>
            <li>
              <b><code>C</code></b>
                  (<code>float</code>, default:
                      <code>1.0</code>
)
              –
              <div class="doc-md-description">
                <p>Regularization strength; must be a positive float. Defaults to 1.0.</p>
              </div>
            </li>
            <li>
              <b><code>return_full</code></b>
                  (<code>bool</code>, default:
                      <code>True</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to return the full classifier object. Defaults to True.</p>
              </div>
            </li>
            <li>
              <b><code>shuffle</code></b>
                  (<code>bool</code>, default:
                      <code>False</code>
)
              –
              <div class="doc-md-description">
                <p>Whether to shuffle the data before splitting. Defaults to False.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>

<table class="field-list">
  <colgroup>
    <col class="field-name" />
    <col class="field-body" />
  </colgroup>
  <tbody valign="top">
    <tr class="field">
      <th class="field-name">Returns:</th>
      <td class="field-body">
        <ul class="first simple">
            <li>
                  <code>(<span title="grnndata.GRNAnnData">GRNAnnData</span>, dict, <span title="sklearn.linear_model.RidgeClassifier">RidgeClassifier</span>)</code>
              –
              <div class="doc-md-description">
                <p>The Gene Regulatory Network data, the metrics of the classifier, and the classifier object.</p>
              </div>
            </li>
        </ul>
      </td>
    </tr>
  </tbody>
</table>
            <details class="quote">
              <summary>Source code in <code>bengrn/base.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">train_classifier</span><span class="p">(</span>
    <span class="n">grn</span><span class="p">:</span> <span class="n">GRNAnnData</span><span class="p">,</span>
    <span class="n">gt</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;omnipath&quot;</span><span class="p">,</span>
    <span class="n">other</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">GRNAnnData</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">use_col</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;symbol&quot;</span><span class="p">,</span>
    <span class="n">train_size</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span>
    <span class="n">doplot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">class_weight</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="p">{</span><span class="mi">1</span><span class="p">:</span> <span class="mi">200</span><span class="p">,</span> <span class="mi">0</span><span class="p">:</span> <span class="mi">1</span><span class="p">},</span>
    <span class="n">max_iter</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1_000</span><span class="p">,</span>
    <span class="n">C</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span>
    <span class="n">return_full</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">shuffle</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    train_classifier trains a classifier to generate a GRN that maps to the ground truth.</span>

<span class="sd">    Uses a RidgeClassifier to select the best combination of networks to predict the ground truth.</span>
<span class="sd">    It is used for the head classification part in the scPRINT paper.</span>

<span class="sd">    Args:</span>
<span class="sd">        grn (GRNAnnData): The Gene Regulatory Network data.</span>
<span class="sd">        gt (str, optional): The name of the ground truth database to use. Defaults to &quot;omnipath&quot;.</span>
<span class="sd">        other (GRNAnnData, optional): Another GRN to compare against. Defaults to None.</span>
<span class="sd">        use_col (str, optional): The column name to use for gene symbols. Defaults to &quot;symbol&quot;.</span>
<span class="sd">        train_size (float, optional): The proportion of the dataset to include in the train split. Defaults to 0.2.</span>
<span class="sd">        doplot (bool, optional): Whether to plot the results. Defaults to True.</span>
<span class="sd">        class_weight (dict, optional): Weights associated with classes in the form {class_label: weight}. Defaults to {1: 200, 0: 1}.</span>
<span class="sd">        max_iter (int, optional): Maximum number of iterations for the classifier. Defaults to 1_000.</span>
<span class="sd">        C (float, optional): Regularization strength; must be a positive float. Defaults to 1.0.</span>
<span class="sd">        return_full (bool, optional): Whether to return the full classifier object. Defaults to True.</span>
<span class="sd">        shuffle (bool, optional): Whether to shuffle the data before splitting. Defaults to False.</span>

<span class="sd">    Returns:</span>
<span class="sd">        (GRNAnnData, dict, RidgeClassifier): The Gene Regulatory Network data, the metrics of the classifier, and the classifier object.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">other</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">elems</span> <span class="o">=</span> <span class="n">other</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">other</span><span class="o">.</span><span class="n">grn</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="n">sub</span> <span class="o">=</span> <span class="n">other</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">grn</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">use_col</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">elems</span><span class="p">)</span><span class="o">.</span><span class="n">targets</span>
        <span class="k">if</span> <span class="n">sub</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">5</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;sub is very small: &quot;</span><span class="p">,</span> <span class="n">sub</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="n">grn</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">use_col</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="n">args</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">genes</span><span class="p">)[</span><span class="n">args</span><span class="p">]</span>
        <span class="n">adj</span> <span class="o">=</span> <span class="n">grn</span><span class="o">.</span><span class="n">varp</span><span class="p">[</span><span class="s2">&quot;GRN&quot;</span><span class="p">][</span><span class="n">args</span><span class="p">,</span> <span class="p">:,</span> <span class="p">:][:,</span> <span class="n">args</span><span class="p">,</span> <span class="p">:][</span><span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">sub</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">)][</span>
            <span class="p">:,</span> <span class="n">np</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="n">sub</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
        <span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;pred shape&quot;</span><span class="p">,</span> <span class="n">adj</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
        <span class="n">da</span> <span class="o">=</span> <span class="n">sub</span><span class="o">.</span><span class="n">values</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">gt</span> <span class="o">=</span> <span class="n">get_GT_db</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">gt</span><span class="p">)</span>
        <span class="n">varnames</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">gt</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">flatten</span><span class="p">())</span>
        <span class="n">intersection</span> <span class="o">=</span> <span class="n">varnames</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">grn</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">use_col</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
        <span class="n">loc</span> <span class="o">=</span> <span class="n">grn</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">use_col</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">intersection</span><span class="p">)</span>
        <span class="n">adj</span> <span class="o">=</span> <span class="n">grn</span><span class="o">.</span><span class="n">varp</span><span class="p">[</span><span class="s2">&quot;GRN&quot;</span><span class="p">][:,</span> <span class="n">loc</span><span class="p">,</span> <span class="p">:][</span><span class="n">loc</span><span class="p">,</span> <span class="p">:,</span> <span class="p">:]</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="n">grn</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">loc</span><span class="p">][</span><span class="n">use_col</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

        <span class="n">da</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">)),</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">gt</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">genes</span> <span class="ow">and</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
                <span class="n">da</span><span class="p">[</span><span class="n">genes</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">i</span><span class="p">),</span> <span class="n">genes</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">j</span><span class="p">)]</span> <span class="o">=</span> <span class="mi">1</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;true elem&quot;</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">da</span><span class="o">.</span><span class="n">sum</span><span class="p">()),</span> <span class="s2">&quot;...&quot;</span><span class="p">)</span>
    <span class="n">da</span> <span class="o">=</span> <span class="n">da</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
    <span class="n">adj</span> <span class="o">=</span> <span class="n">adj</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">adj</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

    <span class="n">X_train</span><span class="p">,</span> <span class="n">X_test</span><span class="p">,</span> <span class="n">y_train</span><span class="p">,</span> <span class="n">y_test</span> <span class="o">=</span> <span class="n">train_test_split</span><span class="p">(</span>
        <span class="n">adj</span><span class="p">,</span> <span class="n">da</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">train_size</span><span class="o">=</span><span class="n">train_size</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="n">shuffle</span>
    <span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;doing classification....&quot;</span><span class="p">)</span>

    <span class="n">clf</span> <span class="o">=</span> <span class="n">RidgeClassifier</span><span class="p">(</span>
        <span class="n">alpha</span><span class="o">=</span><span class="n">C</span><span class="p">,</span>
        <span class="n">fit_intercept</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">class_weight</span><span class="o">=</span><span class="n">class_weight</span><span class="p">,</span>
        <span class="c1"># solver=&quot;saga&quot;,</span>
        <span class="n">max_iter</span><span class="o">=</span><span class="n">max_iter</span><span class="p">,</span>
        <span class="n">positive</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="c1"># clf = LogisticRegression(</span>
    <span class="c1">#    penalty=&quot;l1&quot;,</span>
    <span class="c1">#    C=C,</span>
    <span class="c1">#    solver=&quot;saga&quot;,</span>
    <span class="c1">#    class_weight=class_weight,</span>
    <span class="c1">#    max_iter=max_iter,</span>
    <span class="c1">#    n_jobs=8,</span>
    <span class="c1">#    fit_intercept=False,</span>
    <span class="c1">#    verbose=10,</span>
    <span class="c1">#    **kwargs,</span>
    <span class="c1"># )</span>
    <span class="n">clf</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">X_train</span><span class="p">,</span> <span class="n">y_train</span><span class="p">)</span>
    <span class="n">pred</span> <span class="o">=</span> <span class="n">clf</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">X_test</span><span class="p">)</span>
    <span class="c1"># epr = compute_epr(clf, X_test, y_test)</span>
    <span class="n">metrics</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;used_heads&quot;</span><span class="p">:</span> <span class="p">(</span><span class="n">clf</span><span class="o">.</span><span class="n">coef_</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(),</span>
        <span class="s2">&quot;precision&quot;</span><span class="p">:</span> <span class="p">(</span><span class="n">pred</span><span class="p">[</span><span class="n">y_test</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">/</span> <span class="p">(</span><span class="n">pred</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">(),</span>
        <span class="s2">&quot;random_precision&quot;</span><span class="p">:</span> <span class="n">y_test</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">y_test</span><span class="p">),</span>
        <span class="s2">&quot;recall&quot;</span><span class="p">:</span> <span class="p">(</span><span class="n">pred</span><span class="p">[</span><span class="n">y_test</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">/</span> <span class="n">y_test</span><span class="o">.</span><span class="n">sum</span><span class="p">(),</span>
        <span class="s2">&quot;predicted_true&quot;</span><span class="p">:</span> <span class="n">pred</span><span class="o">.</span><span class="n">sum</span><span class="p">(),</span>
        <span class="s2">&quot;number_of_true&quot;</span><span class="p">:</span> <span class="n">y_test</span><span class="o">.</span><span class="n">sum</span><span class="p">(),</span>
        <span class="c1"># &quot;epr&quot;: epr,</span>
    <span class="p">}</span>
    <span class="k">if</span> <span class="n">doplot</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;metrics&quot;</span><span class="p">,</span> <span class="n">metrics</span><span class="p">)</span>
        <span class="n">PrecisionRecallDisplay</span><span class="o">.</span><span class="n">from_estimator</span><span class="p">(</span>
            <span class="n">clf</span><span class="p">,</span> <span class="n">X_test</span><span class="p">,</span> <span class="n">y_test</span><span class="p">,</span> <span class="n">plot_chance_level</span><span class="o">=</span><span class="kc">True</span>
        <span class="p">)</span>
        <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">return_full</span><span class="p">:</span>
        <span class="n">adj</span> <span class="o">=</span> <span class="n">grn</span><span class="o">.</span><span class="n">varp</span><span class="p">[</span><span class="s2">&quot;GRN&quot;</span><span class="p">]</span>
        <span class="n">grn</span><span class="o">.</span><span class="n">varp</span><span class="p">[</span><span class="s2">&quot;classified&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">clf</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">adj</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">adj</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span>
            <span class="nb">len</span><span class="p">(</span><span class="n">grn</span><span class="o">.</span><span class="n">var</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">grn</span><span class="o">.</span><span class="n">var</span><span class="p">)</span>
        <span class="p">)</span>
    <span class="k">return</span> <span class="n">grn</span><span class="p">,</span> <span class="n">metrics</span><span class="p">,</span> <span class="n">clf</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href=".." class="btn btn-neutral float-left" title="Home"><span class="icon icon-circle-arrow-left"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href=".." style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../search/main.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
